### PR TITLE
Build the "Add a placement" journey for school users

### PR DESCRIPTION
**Pull request by Kizr**
## Context

Users who are part of schools need to be able to add placements to the service.

## Changes proposed in this pull request

This PR mainly adds new logic to handle the multi-page creation of a placement.

- New build model for placement
- New build controller model for palcement
- New views for each page
- Feature spec
- Unit tests

## Guidance to review

Go through the placement creation flow and make sure everything works as expected.

## Link to Trello card

[Build the "Add a placement" journey for school users](https://trello.com/c/ryoxvses)

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated or [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

## Screenshots

image
image
image
image
image
image
image
image

### CODE REVIEW

File: app/models/placements/mentor.rb [hunk 1]
```diff
@@ -21,4 +21,8 @@ class Placements::Mentor < Mentor
   has_many :placements, through: :placement_mentor_joins
 
   default_scope { joins(:mentor_memberships).distinct }
+
```
Comment (JamieCleare2525) [inline app/models/placements/mentor.rb:24]: You shouldn't been this.
https://github.com/DFE-Digital/itt-mentor-services/blob/a4022ef4a7c5ab12c3935e8d9e18be846d261fa9/app/models/mentor.rb#L29

File: app/views/placements/schools/placements/build/_header_content.html.erb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+<%= content_for :page_title, sanitize("Add placement") %>
```
Comment (JamieCleare2525) [inline app/views/placements/schools/placements/build/_header_content.html.erb:1]: I don't think this is used anymore. Can it be removed?

**Review by JamieCleare2525 — COMMENTED** [review]
Error occurs on the mentor page when selecting "Not known"
screencapture-placements-localhost-3000-schools-000251c9-ab23-47b0-9aa7-bca66da457b7-placements-add-mentors-build-check-your-answers-2024-04-15-14_02_05

File: app/controllers/placements/schools/placements/build_controller.rb [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    placement ||= @placement
+    session[:add_a_placement]["mentor_ids"].compact_blank.each do |mentor_id|
+      placement.mentors << Placements::Mentor.find(mentor_id)
+    end
+  end
+
+  def build_subjects(placement = nil)
+    placement ||= @placement
+    subject_ids = session.dig(:add_a_placement, "subject_ids")
+    if subject_ids.instance_of?(String)
```
Comment (JamieCleare2525) [inline app/controllers/placements/schools/placements/build_controller.rb:155]: `placement.subjects << Subject.find(subject_ids)` can handle both the string and the array

Comment (JamieCleare2525) [reply app/controllers/placements/schools/placements/build_controller.rb:155]:
```suggestion
   placement.subjects << Subject.find(subject_ids)
```
^^ This should be enough

File: app/controllers/placements/schools/placements/build_controller.rb [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  def initialize_placement
+    placement = Placements::Schools::Placements::Build::Placement.new(school:, status: :draft)
+    build_mentors(placement)
+    build_subjects(placement)
+    placement
+  end
+
+  def build_mentors(placement = nil)
+    placement ||= @placement
+    session[:add_a_placement]["mentor_ids"].compact_blank.each do |mentor_id|
```
Comment (JamieCleare2525) [inline app/controllers/placements/schools/placements/build_controller.rb:147]:
```suggestion
mentor_ids = session[:add_a_placement]["mentor_ids"].compact_blank
    placement.mentors << Placements::Mentor.find(mentor_ids)
```

File: app/controllers/placements/schools/placements/build_controller.rb [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+  def new
+    reset_session
+    @placement = build_placement
+    school.primary_or_secondary_only? ? handle_primary_or_secondary : handle_other
+  end
+
+  def add_phase
+    @placement = build_placement
+    @selected_phase = session[:add_a_placement]["phase"] || school.phase
```
Comment (JamieCleare2525) [inline app/controllers/placements/schools/placements/build_controller.rb:12]:
```suggestion
    @selected_phase = session.dig(:add_a_placement, "phase") || school.phase
```

File: app/controllers/placements/schools/placements/build_controller.rb [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+  def add_mentors
+    @placement = build_placement
+    @selected_mentors = find_mentors
+  end
+
+  def check_your_answers
+    session[:add_a_placement][:enable_phase_navigation] = true
+    @placement = initialize_placement
+    @phase = session[:add_a_placement]["phase"]
```
Comment (JamieCleare2525) [inline app/controllers/placements/schools/placements/build_controller.rb:29]:
```suggestion
    @phase = session.dig(:add_a_placement, "phase")
```

File: app/controllers/placements/schools/placements/build_controller.rb [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+  end
+
+  def update
+    case params[:id].to_sym
+    when :check_your_answers
+      @placement = Placements::Schools::Placements::Build::Placement.new(school:, phase: session[:add_a_placement]["phase"], status: :published)
```
Comment (JamieCleare2525) [inline app/controllers/placements/schools/placements/build_controller.rb:35]:
```suggestion
      @placement = Placements::Schools::Placements::Build::Placement.new(school:, phase: session.dig(:add_a_placement, "phase"), status: :published)
```

File: app/controllers/placements/schools/placements/build_controller.rb [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+      @placement.mentor_ids = session[:add_a_placement]["mentor_ids"]
```
Comment (JamieCleare2525) [inline app/controllers/placements/schools/placements/build_controller.rb:36]:
```suggestion
      @placement.mentor_ids = session.dig(:add_a_placement, "mentor_ids")
```

File: app/controllers/placements/schools/placements/build_controller.rb [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+      @placement.subject_ids = session[:add_a_placement]["subject_ids"]
```
Comment (JamieCleare2525) [inline app/controllers/placements/schools/placements/build_controller.rb:37]:
```suggestion
      @placement.subject_ids = session.dig(:add_a_placement, 
 "subject_ids")
```

File: app/controllers/placements/schools/placements/build_controller.rb [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+  end
+
+  def assign_subjects_based_on_phase
+    @phase = session.dig(:add_a_placement, "phase").presence || (school.primary_or_secondary_only? ? school.phase : "Primary")
+    @subjects = @phase == "Primary" ? Subject.primary : Subject.secondary
+    @selected_subjects = find_subjects
+  end
+
+  def find_mentors
+    session[:add_a_placement]["mentor_ids"].present? ? school.mentors.find(session.dig(:add_a_placement, "mentor_ids").compact_blank) : []
```
Comment (JamieCleare2525) [inline app/controllers/placements/schools/placements/build_controller.rb:126]:
```suggestion
school.mentors.where(id: session.dig(:add_a_placement, "mentor_ids").compact_blank)
```

File: app/models/placements/schools/placements/build/placement.rb [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  belongs_to :school, class_name: "Placements::School"
+
+  validates :school, :status, presence: true
+
+  attr_accessor :phase, :mentor_ids, :subject_ids
+
+  def valid_phase?
+    return true if phase.present? && [Placements::School::PRIMARY_PHASE, Placements::School::SECONDARY_PHASE].include?(phase)
+
+    errors.add(:phase, "Select a phase")
```
Comment (JamieCleare2525) [inline app/models/placements/schools/placements/build/placement.rb:17]: Define the error text in the locales.

File: app/views/placements/schools/placements/build/add_mentors.html.erb [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+          <% @school.mentors.each do |mentor| %>
+            <%= f.govuk_check_box :mentor_ids, mentor.id, label: { text: mentor.full_name }, checked: @selected_mentors&.include?(mentor) %>
+          <% end %>
+          <%= f.govuk_check_box_divider %>
+          <%= f.govuk_check_box :mentor_ids, :not_known, label: { text: t(".not_known") }, exclusive: true %>
+        <% end %>
+
+        <%= f.govuk_submit t(".continue") %>
+
+        <p class="govuk-body">
```
Comment (JamieCleare2525) [inline app/views/placements/schools/placements/build/add_mentors.html.erb:28]: (Not a blocker) Move cancel link out of the form.

File: config/locales/en/placements/schools/placements.yml [hunk 13]
```diff
@@ -32,3 +72,4 @@ en:
           remove_placement: Remove placement
         destroy:
           placement_removed: Placement removed
+
```
Comment (JamieCleare2525) [inline config/locales/en/placements/schools/placements.yml:75]: `bin/lint` should remove this

File: spec/models/placements/mentor_spec.rb [hunk 14]
```diff
@@ -44,4 +44,11 @@
       expect(described_class.all).not_to include(claims_mentor)
     end
   end
+
+  describe "#full_name" do
```
Comment (JamieCleare2525) [inline spec/models/placements/mentor_spec.rb:48]: This test should be unnecessary given https://github.com/DFE-Digital/itt-mentor-services/blob/32fc251c93508bed6a4706859c0982ae3e90a166/spec/models/mentor_spec.rb#L63

File: spec/system/placements/schools/placements/add_a_placement_spec.rb [hunk 15]
```diff
@@ -0,0 +1,300 @@
+require "rails_helper"
+
+RSpec.describe "Placements / Schools / Placements / Add a placement",
+               type: :system, service: :placements do
+  let(:school) { build(:placements_school, name: "School 1", phase: "Primary") }
+  let(:subject_1) { create(:subject, name: "Primary subject", subject_area: :primary) }
+  let(:subject_2) { create(:subject, name: "Secondary subject", subject_area: :secondary) }
+  let(:mentor_1) { create(:placements_mentor_membership, school:).mentor }
```
Comment (JamieCleare2525) [inline spec/system/placements/schools/placements/add_a_placement_spec.rb:8]: many of these can be `let!` so you don't need them in the `before` block

File: spec/system/placements/schools/placements/add_a_placement_spec.rb [hunk 16]
```diff
@@ -0,0 +1,302 @@
+require "rails_helper"
+
+RSpec.describe "Placements / Schools / Placements / Add a placement",
```
Comment (JamieCleare2525) [inline spec/system/placements/schools/placements/add_a_placement_spec.rb:3]: no test for selecting `not known yet` on mentor

File: app/controllers/placements/schools/placements/build_controller.rb (hunk 6 above)
Comment (JamieCleare2525) [reply app/controllers/placements/schools/placements/build_controller.rb:29]: You marked this as resolved, but I don't think it's changed.

File: app/models/placements/schools/placements/build/placement.rb [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+  belongs_to :school, class_name: "Placements::School"
+
+  validates :school, :status, presence: true
+
+  attr_accessor :phase, :mentor_ids, :subject_ids
+
+  def valid_phase?
+    return true if phase.present? && [Placements::School::PRIMARY_PHASE, Placements::School::SECONDARY_PHASE].include?(phase)
+
+    errors.add(:phase, I18n.t("activerecord.errors.models.placements/schools/placements/build/placement.attributes.phase.invalid"))
```
Comment (JamieCleare2525) [inline app/models/placements/schools/placements/build/placement.rb:17]:
```suggestion
    errors.add(:phase, :invalid))
```

File: app/models/placements/schools/placements/build/placement.rb [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    false
+  end
+
+  def valid_mentor_ids?
+    return true if mentor_ids.present? && mentor_ids.include?("not_known")
+
+    if mentor_ids.present? && mentor_ids.all? { |id| Placements::Mentor.exists?(id:) && school.mentors.exists?(id:) }
+      true
+    else
+      errors.add(:mentor_ids, I18n.t("activerecord.errors.models.placements/schools/placements/build/placement.attributes.mentor_ids.invalid"))
```
Comment (JamieCleare2525) [inline app/models/placements/schools/placements/build/placement.rb:27]:
```suggestion
      errors.add(:mentor_ids, :invalid"))
```

File: app/models/placements/schools/placements/build/placement.rb [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+      false
+    end
+  end
+
+  def valid_subjects?
+    converted_subject_ids = subject_ids.is_a?(Array) ? subject_ids : [subject_ids]
+    if subject_ids.present? && converted_subject_ids.all? { |id| Subject.exists?(id:) && Subject.find(id).subject_area.downcase == phase.downcase }
+      true
+    else
+      errors.add(:subject_ids, I18n.t("activerecord.errors.models.placements/schools/placements/build/placement.attributes.subject_ids.invalid"))
```
Comment (JamieCleare2525) [inline app/models/placements/schools/placements/build/placement.rb:37]:
```suggestion
      errors.add(:subject_ids, :invalid"))
```

**Review by JamieCleare2525 — COMMENTED** [review]
There appear to still be issues with selecting "Not known yet"

https://github.com/DFE-Digital/itt-mentor-services/assets/17162488/833d2036-68cb-455f-84b8-8e69f0321382

**Comment by Kizr** [comment]
> There appear to still be issues with selecting "Not known yet"
> 
>  Screen.Recording.2024-04-16.at.09.12.14.mov

Should be resolved in the latest version

File: app/controllers/placements/schools/placements/build_controller.rb (hunk 6 above)
Comment (Kizr) [reply app/controllers/placements/schools/placements/build_controller.rb]: You're right, sorry about that!

File: config/locales/en/placements/schools/placements.yml (hunk 13 above)
Comment (Kizr) [reply config/locales/en/placements/schools/placements.yml]: Weirdly it did not, I have manually corrected.

File: app/models/placements/schools/placements/build/placement.rb [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+  has_many :subjects, through: :placement_subject_joins
+
+  belongs_to :school, class_name: "Placements::School"
+
+  validates :school, :status, presence: true
+
+  attr_accessor :phase, :mentor_ids, :subject_ids
+
+  def valid_phase?
+    return true if phase.present? && [Placements::School::PRIMARY_PHASE, Placements::School::SECONDARY_PHASE].include?(phase)
```
Comment (JamieCleare2525) [inline app/models/placements/schools/placements/build/placement.rb:15]: Not sure you need the `phase.present?` as `[Placements::School::PRIMARY_PHASE, Placements::School::SECONDARY_PHASE].include?(nil)` will return false

File: app/views/placements/schools/placements/build/check_your_answers.html.erb [hunk 21]
```diff
@@ -0,0 +1,50 @@
+<%= content_for :page_title, sanitize(t(".title")) %>
+<%= render "placements/schools/primary_navigation", school: @school, current_navigation: :placements %>
+
```
Comment (JamieCleare2525) [inline app/views/placements/schools/placements/build/check_your_answers.html.erb:3]: You're missing a back link.
Prototype:
https://school-placements-beta-dev-73591d8bbe5f.herokuapp.com/organisations/049fc44e-6876-4884-8a77-d363173ce8d2/placements/new/check
screencapture-school-placements-beta-dev-73591d8bbe5f-herokuapp-organisations-049fc44e-6876-4884-8a77-d363173ce8d2-placements-new-check-2024-04-17-13_28_04

File: config/locales/en/placements/schools/placements.yml [hunk 22]
```diff
@@ -1,7 +1,58 @@
...
+          add_mentors:
+            title: Mentors
+            add_placement: Add placement
+            mentor: Mentor
+            continue: Continue
+            cancel: Cancel
+            mentors: Mentors
+            not_known: Not known yet
+          check_your_answers:
+            title: Check your answers
```
Comment (JamieCleare2525) [inline config/locales/en/placements/schools/placements.yml:41]:
```suggestion
            title: Check your answers - Add placement
```

File: config/locales/en/placements/schools/placements.yml [hunk 23]
```diff
@@ -1,7 +1,58 @@
...
+            continue: Continue
+            cancel: Cancel
+          add_subject:
+            title: Subject
+            add_placement: Add placement
+            subject: Subject
+            continue: Continue
+            cancel: Cancel
+          add_mentors:
+            title: Mentors
```
Comment (JamieCleare2525) [inline config/locales/en/placements/schools/placements.yml:33]:
```suggestion
            title: Mentors - Add placement
```

File: app/views/placements/schools/placements/build/add_phase.html.erb [hunk 24]
```diff
@@ -0,0 +1,28 @@
+<%= content_for :page_title, sanitize(t(".title")) %>
```
Comment (JamieCleare2525) [inline app/views/placements/schools/placements/build/add_phase.html.erb:1]: https://github.com/DFE-Digital/itt-mentor-services/pull/464/files#r1568769490

File: app/views/placements/schools/placements/build/add_subject.html.erb [hunk 25]
```diff
@@ -0,0 +1,42 @@
+<%= content_for :page_title, sanitize(t(".title")) %>
```
Comment (JamieCleare2525) [inline app/views/placements/schools/placements/build/add_subject.html.erb:1]: https://github.com/DFE-Digital/itt-mentor-services/pull/464/files#r1568769490

**Review by JamieCleare2525 — APPROVED** [review]
Tested locally. Everything appears to be working as requested.

**Pull request merged**